File: client-api/src/test/java/org/ostelco/topup/api/resources/SubscriptionResourceTest.java
package org.ostelco.topup.api.resources;

import io.dropwizard.auth.AuthValueFactoryProvider;
import io.dropwizard.testing.junit.ResourceTestRule;
import io.vavr.control.Either;
import org.glassfish.jersey.test.grizzly.GrizzlyWebTestContainerFactory;
import org.junit.ClassRule;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.ostelco.prime.client.api.model.SubscriptionStatus;
import org.ostelco.prime.model.Price;
import org.ostelco.prime.model.Product;
import org.ostelco.topup.api.auth.AccessTokenPrincipal;
import org.ostelco.topup.api.db.SubscriberDAO;
import org.ostelco.topup.api.util.AccessToken;
import org.ostelco.topup.api.util.AuthDynamicFeatureFactory;

import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.List;

import static java.util.Collections.emptyMap;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class SubscriptionResourceTest {

    private static final SubscriberDAO DAO = mock(SubscriberDAO.class);

    private static final String key = "secret";

    private final String email = "mw@internet.org";

    private final List<Product> acceptedProducts = io.vavr.collection.List.of(
            new Product("1", new Price(10, "NOK"), emptyMap(), emptyMap()),
            new Product("2", new Price(5, "NOK"), emptyMap(), emptyMap()),
            new Product("3", new Price(20, "NOK"), emptyMap(), emptyMap()))
            .toJavaList();
    private final SubscriptionStatus subscriptionStatus = new SubscriptionStatus(5, acceptedProducts);

    @ClassRule
    public static final ResourceTestRule RULE = ResourceTestRule.builder()
            .addResource(AuthDynamicFeatureFactory.createInstance(key))
            .addResource(new AuthValueFactoryProvider.Binder<>(AccessTokenPrincipal.class))
            .addResource(new SubscriptionResource(DAO))
            .setTestContainerFactory(new GrizzlyWebTestContainerFactory())
            .build();

    @Test
    public void getSubscriptionStatus() throws Exception {
        ArgumentCaptor<String> arg = ArgumentCaptor.forClass(String.class);

        when(DAO.getSubscriptionStatus(arg.capture())).thenReturn(Either.right(subscriptionStatus));

        Response resp = RULE.target("/subscription/status")
                .request()
                .header("Authorization", String.format("Bearer %s", AccessToken.withEmail(email)))
                .get(Response.class);

        assertThat(resp.getStatus()).isEqualTo(Response.Status.OK.getStatusCode());
        assertThat(resp.getMediaType().toString()).isEqualTo(MediaType.APPLICATION_JSON);

        // assertThat and assertEquals is not working
        assertTrue(subscriptionStatus.equals(resp.readEntity(SubscriptionStatus.class)));
        assertThat(arg.getValue()).isEqualTo(email);
    }
}

File: client-api/src/test/kotlin/org/ostelco/prime/client/api/resources/SubscriptionResourceTest.kt
package org.ostelco.prime.client.api.resources

import arrow.core.Either
import com.nhaarman.mockito_kotlin.argumentCaptor
import io.dropwizard.auth.AuthDynamicFeature
import io.dropwizard.auth.AuthValueFactoryProvider
import io.dropwizard.auth.oauth.OAuthCredentialAuthFilter
import io.dropwizard.testing.junit.ResourceTestRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.ClassRule
import org.junit.Ignore
import org.junit.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.ostelco.prime.auth.AccessTokenPrincipal
import org.ostelco.prime.auth.OAuthAuthenticator
import org.ostelco.prime.client.api.store.SubscriberDAO
import org.ostelco.prime.client.api.util.AccessToken
import org.ostelco.prime.jsonmapper.objectMapper
import org.ostelco.prime.model.*
import java.time.Instant
import java.util.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

/**
 * Subscription API tests.
 *
 */
class SubscriptionResourceTest {

    private val email = "mw@internet.org"

    private val purchaseRecords = listOf(
            PurchaseRecord(
                    product = Product(sku = "1", price = Price(10, "NOK")),
                    timestamp = Instant.now().toEpochMilli(),
                    id = UUID.randomUUID().toString(),
                    msisdn = ""))

    @Before
    fun setUp() {
        `when`(AUTHENTICATOR.authenticate(ArgumentMatchers.anyString()))
                .thenReturn(Optional.of(AccessTokenPrincipal(email, "email")))
    }

    @Ignore // XXX Fix this asap!
    @Test
    fun getActivePseudonyms() {
        val arg = argumentCaptor<Identity>()

        val msisdn = "4790300001"
        val pseudonym = PseudonymEntity(msisdn, "random", 0, 1)
        val activePseudonyms = ActivePseudonyms(pseudonym, pseudonym)

        `when`(DAO.getActivePseudonymForSubscriber(arg.capture()))
                .thenReturn(Either.right(activePseudonyms))

        val responseJsonString = objectMapper.writeValueAsString(activePseudonyms)

        val resp = RULE.target("/subscription/activePseudonyms")
                .request()
                .header("Authorization", "Bearer ${AccessToken.withEmail(email)}")
                .get(Response::class.java)

        assertThat(resp.status).isEqualTo(Response.Status.OK.statusCode)
        assertThat(resp.mediaType.toString()).isEqualTo(MediaType.APPLICATION_JSON)
        assertTrue(resp.hasEntity())
        assertTrue(responseJsonString == resp.readEntity(String::class.java))
    }

    companion object {

        val DAO: SubscriberDAO = mock(SubscriberDAO::class.java)
        val AUTHENTICATOR: OAuthAuthenticator = mock(OAuthAuthenticator::class.java)

        @JvmField
        @ClassRule
        val RULE: ResourceTestRule = ResourceTestRule.builder()
                .setMapper(objectMapper)
                .addResource(AuthDynamicFeature(
                        OAuthCredentialAuthFilter.Builder<AccessTokenPrincipal>()
                                .setAuthenticator(AUTHENTICATOR)
                                .setPrefix("Bearer")
                                .buildAuthFilter()))
                .addResource(AuthValueFactoryProvider.Binder(AccessTokenPrincipal::class.java))
                .addResource(CustomerResource(DAO))
                .build()
    }
}
